File: core/src/test/java/org/granchi/hollywood/ModelTest.java
package org.granchi.hollywood;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import static com.google.common.truth.Truth.assertThat;

@RunWith(MockitoJUnitRunner.class)
public class ModelTest {
    @Test
    public void testGetSubmodelsOfSameType() throws Exception {
        ParentModel model = new ParentModel();

        assertThat(model.getSubmodelsOfType(ParentModel.class)).containsExactly(model);
    }

    @Test
    public void testGetSubmodelsOfParentType() throws Exception {
        ChildModel model = new ChildModel();

        assertThat(model.getSubmodelsOfType(ParentModel.class)).containsExactly(model);
    }

    @Test
    public void testGetSubmodelsOfChildType() throws Exception {
        ParentModel model = new ParentModel();

        assertThat(model.getSubmodelsOfType(ChildModel.class)).isEmpty();
    }

    @Test
    public void testGetSubmodelsOfUnrelatedType() throws Exception {
        ParentModel model = new ParentModel();

        assertThat(model.getSubmodelsOfType(OtherParentModel.class)).isEmpty();
    }

    private class ParentModel extends Model {
        @Override
        protected Model actUpon(Action action) {
            return null;
        }
    }

    private class ChildModel extends ParentModel {
        @Override
        protected Model actUpon(Action action) {
            return null;
        }
    }

    public class OtherParentModel extends Model {
        @Override
        protected Model actUpon(Action action) {
            return null;
        }
    }
}

File: core/src/test/kotlin/org/granchi/hollywood/ModelTest.kt
package org.granchi.hollywood

import com.google.common.truth.Truth.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class ModelTest {
    @Test
    fun testGetSubmodelsOfSameType() {
        val model = ParentModel()

        assertThat(model.getSubmodelsOfType(ParentModel::class)).containsExactly(model)
    }

    @Test
    fun testGetSubmodelsOfParentType() {
        val model = ChildModel()

        assertThat(model.getSubmodelsOfType(ParentModel::class)).containsExactly(model)
    }

    @Test
    fun testGetSubmodelsOfChildType() {
        val model = ParentModel()

        assertThat(model.getSubmodelsOfType(ChildModel::class)).isEmpty()
    }

    @Test
    fun testGetSubmodelsOfUnrelatedType() {
        val model = ParentModel()

        assertThat(model.getSubmodelsOfType(OtherParentModel::class)).isEmpty()
    }

    private open class ParentModel : Model() {
        override fun actUpon(action: Action): Model? {
            TODO("not implemented")
        }
    }

    private class ChildModel : ParentModel()

    class OtherParentModel : Model() {
        override fun actUpon(action: Action): Model? {
            TODO("not implemented")
        }
    }
}